consensus/src/util/time_service.rs

```diff
@@ -81,32 +82,36 @@ where
 
 /// TimeService implementation that uses actual clock to schedule tasks
 pub struct ClockTimeService {
-    executor: Handle,
+    _handle: thread::JoinHandle<()>,
+    tx: mpsc::SyncSender<(Duration, Box<dyn ScheduledTask>)>,
 }
 
 impl ClockTimeService {
-    /// Creates new TimeService that runs tasks based on actual clock
-    /// It needs executor to schedule internal tasks that facilitates it's work
-    pub fn new(executor: Handle) -> ClockTimeService {
-        ClockTimeService { executor }
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::sync_channel::<(Duration, Box<dyn ScheduledTask>)>(100);
+        let _handle = thread::spawn(move || {
+            while let Ok((timeout, mut task)) = rx.recv() {
+                thread::sleep(timeout);
+                block_on(task.run());
+            }
+        });
+        Self { _handle, tx }
```

Comment:
This only works if you ever have a single task. Is this a safe assumption?

I think the safer way to write it is to use your recv and put the timeout there. If you recv, you add it to a minheap. The minheap tells you what the appropriate timeout should be.

Reply:
it's a safe assumption for now, we can make it fancier later when we need to

Reply:
Let's add a note here then. We may borrow or move this code into secure for key management purposes. cc: @JoshLind